lslforge/haskell/src/Language/Lsl/Render.hs

```diff
@@ -233,4 +238,83 @@ renderPreText :: (Maybe SourceContext) -> String -> String
 renderPreText = maybe (renderString "\n") (renderString . srcPreText)
 
 renderPreText1 :: (String -> String) -> (Maybe SourceContext) -> String -> String
-renderPreText1 f = maybe (renderString "\n" . f) (renderString . srcPreText)
\ No newline at end of file
+renderPreText1 f = maybe (renderString "\n" . f) (renderString . srcPreText)
+
+-- Wrap with parentheses if lower precedence
+
+renderInParenIfLower :: (Ctx Expr) -> ((Ctx Expr) -> Bool) -> String -> String
+renderInParenIfLower ce f =
+        if f ce then renderChar '(' . renderCtxExpr ce . renderChar ')'
+                else renderCtxExpr ce
+
+needsBooleanParens :: Expr -> Ctx Expr -> Bool
+needsBooleanParens ex0 (Ctx _ ex1) =
+        case ex0 of
+            (And _ _) -> case ex1 of
+                                (Or _ _) -> True
+                                _ -> False
+            (Or _ _) -> case ex1 of
+                                (And _ _) -> True
+                                _ -> False
+            _ -> False
+
+castCast :: Expr -> Ctx Expr -> Bool
+castCast ex0 (Ctx _ ex1) =
+        case ex0 of
+            (Cast _ _) -> case ex1 of
+                               (Cast _ _) -> True
+                               _ -> False
+            _ -> False
+
+-- Comparing Order of Precedence
+
+isLower :: Expr -> (Ctx Expr) -> Bool
+isLower e0 (Ctx _ e1) = prec e0 < prec e1
+
+-- Smaller number is higher precedence
+
+prec :: Expr -> Int
+prec e = 
+        case e of
+            (IntLit _) -> 0
+            (FloatLit _) -> 0
+            (StringLit _) -> 0
+            (ListExpr _) -> 0
+            (VecExpr _ _ _) -> 0
+            (RotExpr _ _ _ _) -> 0
+            (KeyLit _) -> 0
+            (Call _ _) -> 0
+            (Cast _ _) -> 1
+            (Not _) -> 2
+            (Inv _) -> 2
+            (Neg _) -> 2
+            (PostInc _) -> 2
+            (PostDec _) -> 2
+            (PreInc _) -> 2
+            (PreDec _) -> 2
+            (Mul _ _) -> 3
+            (Div _ _) -> 3
+            (Mod _ _) -> 3
+            (Add _ _) -> 4
+            (Sub _ _) -> 4
```

Comment:
Have no clue about the Haskell code - but could you test
_1 - 1*(5 - 4)_
as I asume optimizer does not recognize multiplication when removing/opimizing mathematical terms  
this commit https://github.com/RayZopf/lslforge-raysilent/commit/d16dc06380e7b86fdd38eed09616045119df42b4
maybe here

https://github.com/RayZopf/lslforge-raysilent/commit/d16dc06380e7b86fdd38eed09616045119df42b4#diff-f3af7582e55ee4f0c7b8b15a86119f84R170

and here
https://github.com/RayZopf/lslforge-raysilent/commit/d16dc06380e7b86fdd38eed09616045119df42b4#diff-f3af7582e55ee4f0c7b8b15a86119f84L207